File: binding/src/charts/libqscatterseries.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qscatterseries.html
pub const qscatterseries = struct {

    /// New constructs a new QScatterSeries object.
    pub fn New() QtC.QScatterSeries {
        return qtc.QScatterSeries_new();
    }


    /// New2 constructs a new QScatterSeries object.
    pub fn New2(parent: ?*anyopaque) QtC.QScatterSeries {
        return qtc.QScatterSeries_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScatterSeries_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscatterseries.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Type(self: ?*anyopaque, ) i32 {
        return qtc.QScatterSeries_Type(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnType(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QScatterSeries_OnType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseType(self: ?*anyopaque, ) i32 {
        return qtc.QScatterSeries_QBaseType(@ptrCast(self));
    }

    pub fn SetPen(self: ?*anyopaque, pen: ?*anyopaque) void {
        qtc.QScatterSeries_SetPen(@ptrCast(self), @ptrCast(pen));
    }

    /// Allows for overriding the related default method
    pub fn OnSetPen(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QScatterSeries_OnSetPen(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetPen(self: ?*anyopaque, pen: ?*anyopaque) void {
        qtc.QScatterSeries_QBaseSetPen(@ptrCast(self), @ptrCast(pen));
    }

    pub fn SetBrush(self: ?*anyopaque, brush: ?*anyopaque) void {
        qtc.QScatterSeries_SetBrush(@ptrCast(self), @ptrCast(brush));
    }

    /// Allows for overriding the related default method
    pub fn OnSetBrush(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QScatterSeries_OnSetBrush(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetBrush(self: ?*anyopaque, brush: ?*anyopaque) void {
        qtc.QScatterSeries_QBaseSetBrush(@ptrCast(self), @ptrCast(brush));
    }

    pub fn Brush(self: ?*anyopaque, ) QtC.QBrush {
        return qtc.QScatterSeries_Brush(@ptrCast(self));
    }

    pub fn SetColor(self: ?*anyopaque, color: ?*anyopaque) void {
        qtc.QScatterSeries_SetColor(@ptrCast(self), @ptrCast(color));
    }

    /// Allows for overriding the related default method
    pub fn OnSetColor(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QScatterSeries_OnSetColor(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetColor(self: ?*anyopaque, color: ?*anyopaque) void {
        qtc.QScatterSeries_QBaseSetColor(@ptrCast(self), @ptrCast(color));
    }

    pub fn Color(self: ?*anyopaque, ) QtC.QColor {
        return qtc.QScatterSeries_Color(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnColor(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QColor) void {
        qtc.QScatterSeries_OnColor(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseColor(self: ?*anyopaque, ) QtC.QColor {
        return qtc.QScatterSeries_QBaseColor(@ptrCast(self));
    }

    pub fn SetBorderColor(self: ?*anyopaque, color: ?*anyopaque) void {
        qtc.QScatterSeries_SetBorderColor(@ptrCast(self), @ptrCast(color));
    }

    pub fn BorderColor(self: ?*anyopaque, ) QtC.QColor {
        return qtc.QScatterSeries_BorderColor(@ptrCast(self));
    }

    pub fn MarkerShape(self: ?*anyopaque, ) i32 {
        return qtc.QScatterSeries_MarkerShape(@ptrCast(self));
    }

    pub fn SetMarkerShape(self: ?*anyopaque, shape: i32) void {
        qtc.QScatterSeries_SetMarkerShape(@ptrCast(self), @intCast(shape));
    }

    pub fn MarkerSize(self: ?*anyopaque, ) f64 {
        return qtc.QScatterSeries_MarkerSize(@ptrCast(self));
    }

    pub fn SetMarkerSize(self: ?*anyopaque, size: f64) void {
        qtc.QScatterSeries_SetMarkerSize(@ptrCast(self), size);
    }

    pub fn ColorChanged(self: ?*anyopaque, color: QtC.QColor) void {
        qtc.QScatterSeries_ColorChanged(@ptrCast(self), @ptrCast(color));
    }

    pub fn OnColorChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, QtC.QColor) callconv(.c) void) void {
        qtc.QScatterSeries_Connect_ColorChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn BorderColorChanged(self: ?*anyopaque, color: QtC.QColor) void {
        qtc.QScatterSeries_BorderColorChanged(@ptrCast(self), @ptrCast(color));
    }

    pub fn OnBorderColorChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, QtC.QColor) callconv(.c) void) void {
        qtc.QScatterSeries_Connect_BorderColorChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MarkerShapeChanged(self: ?*anyopaque, shape: i32) void {
        qtc.QScatterSeries_MarkerShapeChanged(@ptrCast(self), @intCast(shape));
    }

    pub fn OnMarkerShapeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QScatterSeries_Connect_MarkerShapeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MarkerSizeChanged(self: ?*anyopaque, size: f64) void {
        qtc.QScatterSeries_MarkerSizeChanged(@ptrCast(self), size);
    }

    pub fn OnMarkerSizeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QScatterSeries_Connect_MarkerSizeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScatterSeries_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscatterseries.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScatterSeries_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscatterseries.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QScatterSeries_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/scatterseries.html#types
pub const enums = struct {
    pub const MarkerShape = enum {
        pub const MarkerShapeCircle: i32 = 0;
        pub const MarkerShapeRectangle: i32 = 1;
        pub const MarkerShapeRotatedRectangle: i32 = 2;
        pub const MarkerShapeTriangle: i32 = 3;
        pub const MarkerShapeStar: i32 = 4;
        pub const MarkerShapePentagon: i32 = 5;
    };

};

File: binding/src/charts/libqscatterseries.cpp
#include <QBrush>
#include <QColor>
#include <QObject>
#include <QPen>
#include <QScatterSeries>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QXYSeries>
#include <qscatterseries.h>
#include "libqscatterseries.h"
#include "libqscatterseries.hxx"

QScatterSeries* QScatterSeries_new() {
	 return new VirtualQScatterSeries();
}

QScatterSeries* QScatterSeries_new2(QObject* parent) {
	 return new VirtualQScatterSeries(parent);
}

libqt_string QScatterSeries_Tr(const char* s) {
	QString _ret = QScatterSeries::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QScatterSeries_Type(const QScatterSeries* self) {
	return self->type();
}

void QScatterSeries_SetPen(QScatterSeries* self, const QPen* pen) {
	self->setPen(*pen);
}

void QScatterSeries_SetBrush(QScatterSeries* self, const QBrush* brush) {
	self->setBrush(*brush);
}

QBrush* QScatterSeries_Brush(const QScatterSeries* self) {
	return new QBrush(self->brush());
}

void QScatterSeries_SetColor(QScatterSeries* self, const QColor* color) {
	self->setColor(*color);
}

QColor* QScatterSeries_Color(const QScatterSeries* self) {
	return new QColor(self->color());
}

void QScatterSeries_SetBorderColor(QScatterSeries* self, const QColor* color) {
	self->setBorderColor(*color);
}

QColor* QScatterSeries_BorderColor(const QScatterSeries* self) {
	return new QColor(self->borderColor());
}

int QScatterSeries_MarkerShape(const QScatterSeries* self) {
	return self->markerShape();
}

void QScatterSeries_SetMarkerShape(QScatterSeries* self, int shape) {
	self->setMarkerShape(static_cast<QScatterSeries::MarkerShape>(shape));
}

double QScatterSeries_MarkerSize(const QScatterSeries* self) {
	return self->markerSize();
}

void QScatterSeries_SetMarkerSize(QScatterSeries* self, double size) {
	self->setMarkerSize(size);
}

void QScatterSeries_ColorChanged(QScatterSeries* self, QColor* color) {
	self->colorChanged(*color);
}

void QScatterSeries_Connect_ColorChanged(QScatterSeries* self, intptr_t slot) {
    void (*slotFunc)(QScatterSeries*, QColor*) = reinterpret_cast<void (*)(QScatterSeries*, QColor*)>(slot);
    QScatterSeries::connect(self, &QScatterSeries::colorChanged, [self, slotFunc](QColor color) {
	slotFunc(self, color);
    });
}

void QScatterSeries_BorderColorChanged(QScatterSeries* self, QColor* color) {
	self->borderColorChanged(*color);
}

void QScatterSeries_Connect_BorderColorChanged(QScatterSeries* self, intptr_t slot) {
    void (*slotFunc)(QScatterSeries*, QColor*) = reinterpret_cast<void (*)(QScatterSeries*, QColor*)>(slot);
    QScatterSeries::connect(self, &QScatterSeries::borderColorChanged, [self, slotFunc](QColor color) {
	slotFunc(self, color);
    });
}

void QScatterSeries_MarkerShapeChanged(QScatterSeries* self, int shape) {
	self->markerShapeChanged(static_cast<QScatterSeries::MarkerShape>(shape));
}

void QScatterSeries_Connect_MarkerShapeChanged(QScatterSeries* self, intptr_t slot) {
    void (*slotFunc)(QScatterSeries*, int) = reinterpret_cast<void (*)(QScatterSeries*, int)>(slot);
    QScatterSeries::connect(self, &QScatterSeries::markerShapeChanged, [self, slotFunc](MarkerShape shape) {
	slotFunc(self, shape);
    });
}

void QScatterSeries_MarkerSizeChanged(QScatterSeries* self, double size) {
	self->markerSizeChanged(size);
}

void QScatterSeries_Connect_MarkerSizeChanged(QScatterSeries* self, intptr_t slot) {
    void (*slotFunc)(QScatterSeries*, double) = reinterpret_cast<void (*)(QScatterSeries*, double)>(slot);
    QScatterSeries::connect(self, &QScatterSeries::markerSizeChanged, [self, slotFunc](qreal size) {
	slotFunc(self, size);
    });
}

libqt_string QScatterSeries_Tr2(const char* s, const char* c) {
	QString _ret = QScatterSeries::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QScatterSeries_Tr3(const char* s, const char* c, int n) {
	QString _ret = QScatterSeries::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
int QScatterSeries_QBaseType(const QScatterSeries* self) {
	auto* vqscatterseries = dynamic_cast<const VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_Type_IsBase(true);
	return vqscatterseries->type();
}
}

// Auxiliary method to allow providing re-implementation
void QScatterSeries_OnType(const QScatterSeries* self, intptr_t slot) {
	auto* vqscatterseries = dynamic_cast<const VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_Type_Callback(reinterpret_cast<VirtualQScatterSeries::QScatterSeries_Type_Callback>(slot));
}
}

// Base class handler implementation
void QScatterSeries_QBaseSetPen(QScatterSeries* self, const QPen* pen) {
	auto* vqscatterseries = dynamic_cast<VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_SetPen_IsBase(true);
	vqscatterseries->setPen(*pen);
}
}

// Auxiliary method to allow providing re-implementation
void QScatterSeries_OnSetPen(QScatterSeries* self, intptr_t slot) {
	auto* vqscatterseries = dynamic_cast<VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_SetPen_Callback(reinterpret_cast<VirtualQScatterSeries::QScatterSeries_SetPen_Callback>(slot));
}
}

// Base class handler implementation
void QScatterSeries_QBaseSetBrush(QScatterSeries* self, const QBrush* brush) {
	auto* vqscatterseries = dynamic_cast<VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_SetBrush_IsBase(true);
	vqscatterseries->setBrush(*brush);
}
}

// Auxiliary method to allow providing re-implementation
void QScatterSeries_OnSetBrush(QScatterSeries* self, intptr_t slot) {
	auto* vqscatterseries = dynamic_cast<VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_SetBrush_Callback(reinterpret_cast<VirtualQScatterSeries::QScatterSeries_SetBrush_Callback>(slot));
}
}

// Base class handler implementation
void QScatterSeries_QBaseSetColor(QScatterSeries* self, const QColor* color) {
	auto* vqscatterseries = dynamic_cast<VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_SetColor_IsBase(true);
	vqscatterseries->setColor(*color);
}
}

// Auxiliary method to allow providing re-implementation
void QScatterSeries_OnSetColor(QScatterSeries* self, intptr_t slot) {
	auto* vqscatterseries = dynamic_cast<VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_SetColor_Callback(reinterpret_cast<VirtualQScatterSeries::QScatterSeries_SetColor_Callback>(slot));
}
}

// Base class handler implementation
QColor* QScatterSeries_QBaseColor(const QScatterSeries* self) {
	auto* vqscatterseries = dynamic_cast<const VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_Color_IsBase(true);
	return new QColor(vqscatterseries->color());
}
}

// Auxiliary method to allow providing re-implementation
void QScatterSeries_OnColor(const QScatterSeries* self, intptr_t slot) {
	auto* vqscatterseries = dynamic_cast<const VirtualQScatterSeries*>(self);
	if (vqscatterseries && vqscatterseries->isVirtualQScatterSeries) {
vqscatterseries->setQScatterSeries_Color_Callback(reinterpret_cast<VirtualQScatterSeries::QScatterSeries_Color_Callback>(slot));
}
}

void QScatterSeries_Delete(QScatterSeries* self) {
    delete self;
}

